docs/pages/blog/react-19-update.md

```diff
@@ -0,0 +1,142 @@
+
+The existence of the `ref` prop on a `ForwardRef` component, even if `undefined`, makes the component props referentially unstable, breaking the downstream memoizations.
+
+To address this, a `forwardRef` shim was added that enforces correct prop ordering at the type level.
+
+```tsx
+// Compatibility shim that ensures stable props object for forwardRef components
+// Fixes https://github.com/facebook/react/issues/31613
+// We ensure that the ref is always present in the props object (even if that's not the case for older versions of React) to avoid the footgun of spreading props over the ref in the newer versions of React.
+export const forwardRef = <T, P = {}>(
+  render: React.ForwardRefRenderFunction<T, P & { ref: React.Ref<T> }>,
+) => {
+  if (reactMajor >= 19) {
+    const Component = (props: any) => render(props, props.ref ?? null);
+    Component.displayName = render.displayName ?? render.name;
+    return Component as React.ForwardRefExoticComponent<P>;
+  }
+  return React.forwardRef(
+    render as React.ForwardRefRenderFunction<T, React.PropsWithoutRef<P>>,
+  );
+};
+```
+
+The shim provides two key benefits:
+
+1. Type safety - TypeScript will warn if props are spread incorrectly
+2. Forward compatibility - Components using the shim will work correctly in all supported React versions
+
+The code update was:
+
+```tsx
+// Before
+const GridRoot = forwardRef((props, ref) => {
+  const state = useGridState();
+  return <div ref={ref} {...props} {...state} />;
+});
+
+// After
+const GridRoot = forwardRefShim((props, ref) => {
+  const state = useGridState();
+  return <div {...props} {...state} ref={ref} />;
+});
+```
+
+## Phase 2: Moving to React 19
+
+After ensuring compatibility, we started working on migrating the codebase to React 19. This involved:
+
+1. Updating all package dependencies to their React 19 compatible versions
+2. Migrating test utilities to work with React 19
+3. Ensuring all components work with the new React 19 features
+4. Updating CI to run tests with React 18
+5. Type references with `RefObject` for React 19 and with `MutableRefObject` for earlier versions.
+
+The biggest change in this phase was around the `useRef()` hook [update](https://react.dev/blog/2024/04/25/react-19-upgrade-guide#useref-requires-argument) and the fact that `apiRef` in the DataGrid component had to be updated from `MutableRefObject` to `RefObject` for React 19 only, in order to avoid type errors for users (and their CI) that did not migrate yet.
+
+As part of the migration we also moved to NextJS 15 for our documentation website.
+
+### Our own `RefObject`
+
+To have different types for `apiRef` in the DataGrid component for different React versions, we created our own `RefObject` type.
+
+We used the fact that `useRef()` requires a param for React 19 to have `RefObject` evaluated to `MutableRefObject` for React < 19 and to `RefObject` otherwise.
```

Comment:
I usually hate the word "leverage" in software writing because it gets way overused but this is a rare case in which I think it's the appropriate word choice. 😁 Aside from that, I found this sentence difficult to parse, so I made an attempt at clarifying it based on my understanding.

```suggestion
We leveraged the fact that `useRef()` requires a param for React 19, and set it up so that `RefObject` would be evaluated to `MutableRefObject` for React <19, or else `RefObject` otherwise:
```

Reply:
"Or else [..] otherwise" might be a bit redundant/a pleonasm? Perhaps something like this:

```suggestion
We leveraged the fact that `useRef()` requires a param in React 19 to ensure `RefObject` would be evaluated as `MutableRefObject` for React < 19, and as `RefObject` otherwise.
```